### PR TITLE
[Serve] Serialize Query object directly

### PR DESCRIPTION
**Pull request by simon-mo**
## Why are these changes needed?
This PR serializes the query object directly rather than go through the ray serialization context. When we serialize it directly, ray recognize the input is now bytes and it will skip the entire cloudpickle stack. This performance issue is identified by looking at the following flamegraph. The lower center/left stack is the serialization stack.

image

benchmark:

- Using config {'num_replicas': 8, 'max_concurrent_queries': 10000}
- Before: 2381 qps
- After: 2718 qps
- 14% increase in queries per second.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: python/ray/serve/router.py [hunk 1]
```diff
@@ -316,13 +316,14 @@ async def _do_query(self, backend, backend_replica_tag, req):
         start = time.time()
         worker = self.replicas[backend_replica_tag]
         try:
+            object_ref = worker.handle_request.remote(req.ray_serialize())
             if req.is_shadow_query:
                 # No need to actually get the result, but we do need to wait
                 # until the call completes to mark the worker idle.
-                asyncio.wait([worker.handle_request.remote(req)])
+                await asyncio.wait([object_ref])
```
Comment (simon-mo) [inline python/ray/serve/router.py:323]: This is a bug before, we should have waited for the object

File: python/ray/serve/tests/test_router.py [hunk 2]
```diff
@@ -204,11 +212,11 @@ def get_queues(self):
     worker = MockWorker.remote()
     q = ray.remote(VisibleRouter).remote()
     await q.setup.remote()
-    BACKEND_NAME = "max-concurrent-test"
+    backend_name = "max-concurrent-test"
```
Comment (simon-mo) [inline python/ray/serve/tests/test_router.py:215]: somehow lint starts complaining about this casing issue.

**Comment by suquark** [comment]
I think for whatever kind of serialization without complex objects (functions, classes, types, ...), we should use pickle directly. Cloudpickle is slow. Also there are some tricks about defining serializers, see #9462

File: python/ray/serve/tests/test_router.py (hunk 2 above)
Comment (edoakes) [reply python/ray/serve/tests/test_router.py:215]: weird

File: python/ray/serve/tests/test_router.py [hunk 3]
```diff
@@ -186,6 +188,12 @@ async def test_shard_key(serve_instance, task_runner_mock_actor):
 
 
 async def test_router_use_max_concurrency(serve_instance):
+    # The VisibleRouter::get_queues method needs to pickle queries
```
Comment (edoakes) [inline python/ray/serve/tests/test_router.py:191]: Like I mentioned in the other PR, can we just remove the async_future member so we don't need these custom serializers?

File: python/ray/serve/router.py (hunk 1 above)
Comment (edoakes) [reply python/ray/serve/router.py:323]: Oh wow, good catch!

**Review by edoakes — COMMENTED** [review]
Interesting, I'm somewhat surprised that just doing the serialization manually improves performance. This indicates that there's some hidden overhead in the ray serialization code to me (cc @suquark).

**Comment by simon-mo** [comment]
@edoakes I'm using a shortcut in serialization to directly skip the entire cloudpicke stack, basically, bytes are directly serialized with very low overhead https://github.com/ray-project/ray/blob/60a838f0f012af4f73ec3e59e5f55adffe4f8178/python/ray/serialization.py#L372-L382

**Comment by simon-mo** [comment]
Thanks @suquark I applied the suggestion here https://github.com/ray-project/ray/commit/60a838f0f012af4f73ec3e59e5f55adffe4f8178

**Review by edoakes — APPROVED** [review]

**Pull request merged**